src/core/types.rs

```diff
@@ -0,0 +1,303 @@
+use std::collections::{HashSet, HashMap, BTreeSet};
+use std::cell::{RefCell, Ref};
+use std::ops::Deref;
+use log::{debug};
+
+use pyo3::prelude::*;
+
+use crate::importlib;
+
+#[pyclass]
+#[derive(Clone)]
+pub struct DepNode {
+    #[pyo3(get)]
+    pub name: String,
+    #[pyo3(get)]
+    pub spec: importlib::ModuleSpec,
+    // The dependencies & dependents by spec.name
+    dependencies: HashSet<String>,
+    #[pyo3(get)]
+    dependents: HashSet<String>,
+    #[pyo3(get)]
+    depth: Option<i32>
+}
+
+impl DepNode {
+    pub fn new(spec: importlib::ModuleSpec, depth: Option<i32>) -> DepNode {
+        DepNode {
+            name: spec.name.clone(),
+            spec: spec,
+            dependencies: HashSet::new(),
+            dependents: HashSet::new(),
+            depth: depth // Allow for uninitialized depths
+        }
+    }
+
+    fn merge(&mut self, other: DepNode) {
+        // Few sanity checks
+        assert!(self.name == other.name);
+        if let Some(origin) = &self.spec.origin {
+            assert!(origin == &other.spec.origin.unwrap());
+        }
+
+        // Merge data
+        self.dependencies.extend(other.dependencies);
+        self.dependents.extend(other.dependents);
+        if other.depth < self.depth {
+            self.depth = other.depth
+        }
+    }
+
+    #[allow(dead_code)]
+    fn is_root(&self) -> bool {
+        self.dependencies.len() == 0
```

Comment:
nit: rust linter (clippy) would prefer `self.dependencies.is_empty()`

Sidenote - probably overkill for this project, but for larger projects I would recommend adding clippy to CI pipelines to ensure linting standard is adhered to